### PR TITLE
Improve mapping immutable objects

### PR DESCRIPTION
**Pull request by zenangst**
This PR tries to address the issue where you want to map properties to an immutable model.

### SafeMappable protocol
```swift
public protocol SafeMappable {
  init(_ map: JSONDictionary) throws
}
```

The new `SafeMappable` protocol can be used for mapping immutable models.

### Usage
```swift
struct TestImmutable: SafeMappable {
  var firstName: String
  var lastName: String

  init(_ map: JSONDictionary) throws {
    firstName = try <-map.property("firstName")
    lastName = try <-map.property("lastName")
  }
}
```

This PR also has some breaking changes; `object` and `objects` have been replaced by `property` and `properties`.
This change was made because I felt that it was more semantically correct.

tl;dr

- [x] Rename `object()` to `property()`
- [x] Rename `objects()` to `properties()`
- [x] Split source in to separate files
- [x] Introduce prefix operator `<-`
- [x] Add `SafeMappable` protocol
- [x] Add `propertyOrThrow` in `Dictionary` extension
- [x] Refactor code style
- [x] Remove inspectable

Tries to fix https://github.com/zenangst/Tailor/issues/23

### CODE REVIEW

File: Cartfile [hunk 1]
```diff
@@ -1 +1 @@
-github "hyperoslo/Sugar" "master"
+github "hyperoslo/Sugar" "feature/result"
```
Comment (zenangst) [inline Cartfile]: This should be changed to `master` before merging.

Comment (zenangst) [inline Cartfile]: Or `tag` for that matter.

File: Sources/Shared/Extensions/Dictionary+Tailor.swift [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    return value as? T
+  }
+
+  func transform<T, U>(name: String, transformer: ((value: U?) -> T?)) -> T? {
+    guard let value = self[name as! Key] else { return nil }
+    return transformer(value: value as? U)
+  }
+
+  func relation<T : Mappable>(name: String) -> T? {
```
Comment (zenangst) [inline Sources/Shared/Extensions/Dictionary+Tailor.swift]: This changed from `object` to `relation`.

File: Sources/Shared/Extensions/Dictionary+Tailor.swift [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    guard let value = self[name as! Key],
+      dictionary = value as? JSONDictionary
+      else { return nil }
+
+    return T(dictionary)
+  }
+
+  func relations<T : Mappable>(name: String) -> [T]? {
```
Comment (zenangst) [inline Sources/Shared/Extensions/Dictionary+Tailor.swift]: This is the function formerly known as `objects`.

File: Sources/Shared/Operators/Operators.swift [hunk 4]
```diff
@@ -0,0 +1,30 @@
+prefix operator <- {}
+infix operator <- {}
+infix operator <+ {}
+
+public prefix func <-<T>(rhs: T?) throws -> T {
+  guard let rhs = rhs else {
+    throw MappableError.TypeError(message: "Unable to unrwap value")
+  }
+  return rhs
+}
```
Comment (zenangst) [inline Sources/Shared/Operators/Operators.swift]: This is the new prefix operator that you can use for mapping immutable models.

File: TailorTests/Shared/TestSubjects.swift [hunk 5]
```diff
@@ -78,8 +78,20 @@ struct TestPersonStruct: Inspectable, Mappable, Equatable {
...
+  }
+}
+
+struct TestImmutable: MappableOnce {
+  var firstName: String
+  var lastName: String
+
+  init(_ map: JSONDictionary) throws {
+    do {
+      firstName = try <-map.property("firstName")
```
Comment (RamonGilabert) [inline TailorTests/Shared/TestSubjects.swift]: Would you add a space here, like: `firstName = try <- map.property("firstName")`?

File: TailorTests/Shared/TestSubjects.swift [hunk 6]
```diff
@@ -78,8 +78,20 @@ struct TestPersonStruct: Inspectable, Mappable, Equatable {
...
 
-    relatives <- map.objects("relatives")
-    job <- map.object("job")
+    relatives <- map.relations("relatives")
+    job <- map.relation("job")
+  }
+}
+
+struct TestImmutable: MappableOnce {
+  var firstName: String
```
Comment (kostiakoval) [inline TailorTests/Shared/TestSubjects.swift]:
``` 
let firstName: String
let lastName: String
```

Comment (zenangst) [inline TailorTests/Shared/TestSubjects.swift]: Oh right, my bad! ^^

**Comment by kostiakoval** [comment]
Use `let`s, It looks like possible :) :+1:  
```
struct TestImmutable: MappableOnce {
  let firstName: String
  let lastName: String
```

File: TailorTests/Shared/TestSubjects.swift (hunk 6 above)
Comment (zenangst) [inline TailorTests/Shared/TestSubjects.swift]: @kostiakoval fixed :sunglasses:

File: TailorTests/Shared/TestSubjects.swift (hunk 5 above)
Comment (zenangst) [inline TailorTests/Shared/TestSubjects.swift]: @RamonGilabert can't because it's a prefix operator, it needs to be attached :sunglasses:

Comment (zenangst) [inline TailorTests/Shared/TestSubjects.swift]: I would if I could

Comment (kostiakoval) [inline TailorTests/Shared/TestSubjects.swift]: Looks like it's not possible right now :(
It's a `prefix func <-` function

**Comment by zenangst** [comment]
@kostiakoval 

screen shot 2016-02-16 at 20 54 24

**Comment by kostiakoval** [comment]
I really like the functionality. :100: 

The only thing that looks a bit scary that you have to use `<-` prefix operator.
There are exist both prefix and infix operators `<-`  It could be confusing for users when to use which.

**Comment by zenangst** [comment]
@kostiakoval I kind like that they are similar but I share your concern that forgetting to add or don't add a space there could cause some weird issues. We could try to find a more appropriate prefix operator or remove it entirely and just use `propertyOrThrow` instead of `property`.

File: Sources/Shared/Protocols/MappableOnce.swift [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import Sugar
+
+public protocol MappableOnce {
+  init?(_ map: JSONDictionary) throws
```
Comment (vadymmarkov) [inline Sources/Shared/Protocols/MappableOnce.swift]: I think you don't need `?` if it throws.

Comment (zenangst) [inline Sources/Shared/Protocols/MappableOnce.swift]: Nice catch!

Comment (vadymmarkov) [inline Sources/Shared/Protocols/MappableOnce.swift]: And could it be just one protocol called `Mappable` that throws.

Comment (zenangst) [inline Sources/Shared/Protocols/MappableOnce.swift]: We could totally do that, but that would be a slightly bigger refactoring as both `relation` and `relations` create `Mappable` objects and those methods also need to handle the exceptions.
Not saying that it's a bad thing and we should totally look into it, however it would work for @kostiakoval needs without that :sunglasses:

File: Sources/Shared/Extensions/Dictionary+Tailor.swift (hunk 2 above)
Comment (vadymmarkov) [inline Sources/Shared/Extensions/Dictionary+Tailor.swift]: @zenangst What about 
`func property<T: Mappable>(name: String) -> T?` 
`func property<T : Mappable>(name: String) -> [T]?`

Like we discussed today, the same name, but for multiple possible types.

File: Sources/Shared/Protocols/MappableOnce.swift (hunk 7 above)
Comment (vadymmarkov) [inline Sources/Shared/Protocols/MappableOnce.swift]: Just think that one protocol is better that 2 for public API, also the name `MappableOnce` is a bit weird, in both cases it's initialiser, so you can map only once :smile:

Comment (zenangst) [inline Sources/Shared/Protocols/MappableOnce.swift]: Very good point, didn't think about that ^^

File: Sources/Shared/Extensions/Dictionary+Tailor.swift (hunk 2 above)
Comment (zenangst) [inline Sources/Shared/Extensions/Dictionary+Tailor.swift]: Tried it but for some reason it won't work with optional values.

Comment (zenangst) [inline Sources/Shared/Extensions/Dictionary+Tailor.swift]: Seems like `relation` and `property` collide.

```swift
Tailor/TailorTests/Shared/TestSubjects.swift:72:15: Ambiguous use of operator '<-'
```

Comment (zenangst) [inline Sources/Shared/Extensions/Dictionary+Tailor.swift]: Oh wait, I know what's up... I think

**Comment by zenangst** [comment]
@kostiakoval what do you think of the latest changes?

**Comment by Danappelxx** [comment]
Completely unrelated but in your example you do:
```swift
init(_ map: JSONDictionary) throws {
    do {
      firstName = try <-map.property("firstName")
      lastName = try <-map.property("lastName")
    }
}
```
when you could (should) do:
```swift
init(_ map: JSONDictionary) throws {
    firstName = try <-map.property("firstName")
    lastName = try <-map.property("lastName")
}
```
(you don't need the `do` because the `throws` does that for you).

**Comment by zenangst** [comment]
@Danappelxx Not completely unrelated imo :sunglasses: 
Thanks for the feedback, should be fixed now :)

**Comment by kostiakoval** [comment]
This looks good to me. I would merge this PR but I think would wait with 1.0.0 maybe.
I have some awesome ideas for improvements.

**Comment by zenangst** [comment]
@kostiakoval @vadymmarkov @Danappelxx @RamonGilabert thanks for all your feedback! <3
